src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+import java.util.*;
+
+public class EnsemblBiomartRetriever extends FileRetriever {
+
+    private Properties properties;
+
+    public EnsemblBiomartRetriever() { }
+
+    public EnsemblBiomartRetriever(String retrieverName)
+    {
+        super(retrieverName);
+    }
+
+    private static final String  microarrayTypesBaseQuery = "type=listAttributes&mart=ENSEMBL_MART_ENSEMBL&virtualSchema=default&dataset=BIOMART_SPECIES_NAME_gene_ensembl&interface=default&attributePage=feature_page&attributeGroup=external&attributeCollection=microarray";
+
+    /**
+     * Downloads Ensembl-Microarray and Ensembl-Uniprot identifier mapping files for all species, if they exist.
+     * @throws IOException - Thrown when file can't be found when during writing and by HTTPConnection class.
+     * @throws ParseException - Thrown by JSONParser when reading Species config file.
+     * @throws InterruptedException - Thrown if Sleep is interrupted when waiting to retry BioMart query.
+     * @throws BioMartQueryException - Thrown if BioMart query doesn't match any existing data in their database.
+     * @throws HttpException - Thrown when the Http request to BioMart returns a non-200 response.
+     */
+    public void downloadData() throws IOException, ParseException, InterruptedException, BioMartQueryException, HttpException {
+
+        // Create directory where Biomart files will be stored.
+        Files.createDirectories(Paths.get(this.destination));
+        setProperties(EnsemblBiomartUtil.getProperties());
+
+        // Get names of all organisms we add links and/or microarray data for.
+        // Species names are in biomart format (eg: hsapiens).
+        for (String speciesName : EnsemblBiomartUtil.getSpeciesNames()) {
+            String speciesBiomartName = EnsemblBiomartUtil.getBiomartSpeciesName(speciesName);
+            logger.info("Retrieving Biomart files for " + speciesBiomartName);
+
+            logger.info("Retrieving microarray data");
+            // Query Biomart for existing microarray 'types' (not ids) that exist for this species.
+            Set<String> microarrayTypes = queryBiomart(getMicroarrayTypesQuery(speciesBiomartName), 0);
+            // Iterate through each microarray type and retrieve Ensembl-Microarray identifier mappings.
+            // All mappings are stored in a single file, (eg: hsapiens_microarray);
+            for (String microarrayType : microarrayTypes) {
+                queryBiomartAndStoreData(speciesBiomartName, getBiomartXMLFilePath(), microarrayType, "_microarray");
+            }
+
+            // Query Ensembl-Uniprot (swissprot and trembl) identifier mapping data from Biomart
+            // and write it to a file (eg: hsapiens_uniprot).
+            logger.info("Retrieving UniProt data");
+            for (String uniprotQueryId : Arrays.asList("uniprotswissprot", "uniprotsptrembl")) {
+                queryBiomartAndStoreData(speciesBiomartName, getBiomartXMLFilePath(), uniprotQueryId, "_uniprot");
```

Comment:
A named constant like `final String uniProtFileSuffix = "_uniprot";`?  There may be enough literals that they could be moved to common place such as the new utlitiies class or a separate constants class.

Reply:
Done. I changed a couple of other values (like 'uniprotswissprot' and 'unitprotsptrembl') to be named constants as well.